### DISCUSSION TITLE
Overriding types returned by prisma queries?

### DISCUSSION THREAD

**Discussion by olalonde** [prisma/prisma#21112] [Q&A]
### Question

I was wondering if there is a way to override the types that are returned by Prisma queries. For example, I have a `Session` Model that can be in different states, e.g. `pending`, `ongoing`, `closed`. Each state guarantees a different set of fields to be not null. For example, the `closed` state guarantees that `closedAt` will not be `null`. This can be easily represented in TypeScript, e.g.:

```typescript
type PendingSession = { type: "pending", startedAt: null, closedAt: null }
type OngoingSession = { type: "ongoing", startedAt: Date, closedAt: null }
type ClosedSession = { type: "closed", startedAt: Date, closedAt: Date }
type Session = PendingSession | OngoingSession | ClosedSession;
```

Now the type that Prisma generates looks more like this:

```typescript
type Session = { startedAt?: Date, closedAt?: Date }
```

I would like to parse the session returned and assign it a more convenient type like the one mentioned above. I could do it manually every time I query a session / sessions, but this adds a lot of boilerplate and I have to remember to do it, including when a session is returned as part of an `include: { session: true }` query (that's really the hard part). 

I was wondering if there's a way to alter the default behavior so that it does it automatically for me without having to do it manually. Can an extension do this and if possible an example would be appreciated.

**Comment by ludralph** [comment]
Hi @olalonde 👋 

Thank you for raising this question.

At this moment, Prisma doesn't have support for overriding the generated types but we do have a similar feature request here:

- https://github.com/prisma/prisma/issues/12567

Kindly document your use case and leave a 👍 to help our engineers prioritize it.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ludralph** [reply]
Hi @olalonde 

I also checked in with the engineering team, and the advice is to implement a type guard as a model extension
https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
